File: src/main/scala/ccampo133/nbody/Body.scala
package ccampo133.nbody

/**
 * @author Chris Campo
 */
class Body(
  val mass: Double,
  val radius: Double,
  val position: Vec2D,
  val velocity: Vec2D,
  var positions: Seq[Vec2D] = Vector.empty[Vec2D]) {

  if (positions.isEmpty) positions :+= position

  def isCollision(other: Body): Boolean = {
    val r = other.position - position
    r.length() < other.radius && mass <= other.mass
  }
}

File: src/main/kotlin/me/ccampo/nbody/model/Body.kt
package me.ccampo.nbody.model

/**
 * Represents a circular moving body in 2D space.
 *
 * @param m The body's mass.
 * @param r The body's radius.
 * @param x The body's position.
 * @param v The body's velocity.
 * @param positions A list of the body's previous positions
 */
data class Body(val m: Double, val r: Double, val x: Vector, val v: Vector, val positions: List<Vector> = emptyList()) {
  /**
   * Returns true if this body has collided (occupies the same space) with another body, otherwise returns false.
   *
   * @param other The other body to detect collisions with.
   */
  fun isCollision(other: Body) = (this.x - other.x).len < other.r && this.m <= other.m

  companion object {
    val empty = Body(0.0, 0.0, Vector.zero, Vector.zero)
  }
}
